### PR TITLE
Implement CKMSSample

### PR DESCRIPTION
**Pull request by hidenori-shinohara**
Use the CKMS quantile approximation algorithm and implement `CKMSSample`. `CKMSSample` maintains two sliding N-second windows. One is the current window (e.g., [11:00:30, 11:01:00] at 11:00:45) and the other one is the previous window (e.g., [11:00:00, 11:00:30] at 11:00:45) for reporting.

The window size is adjustable (by default N = 30 seconds), and every time window's starting time is a multiple of `N` for consistency. (i.e., `startingTime.seconds_since_epoch() % N == 0`) This uses the previous window for reporting so that we know that the metrics are always based on N-seconds' of data. Furthermore, the `Histogram` class will report metrics based on the same time window when using `CKMSSample`. For instance, the maximum value will be based on the previous `N`-second window, rather than the all-time maximum values.

### CODE REVIEW

File: src/medida/stats/ckms_sample.cc [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+}
+
+// === Implementation ===
+
+Clock::time_point CKMSSample::Impl::CalculateCurrentWindowStartingPoint(Clock::time_point time) const {
+    return time - (std::chrono::duration_cast<std::chrono::seconds>(time.time_since_epoch()) % window_size_);
+}
+
+void CKMSSample::Impl::AssertValidTime(Clock::time_point const& timestamp) const {
+    assert(last_asserted_time_ <= timestamp);
```
Comment (graydon) [inline src/medida/stats/ckms_sample.cc]: This is a little more strict than I think it ought to be: the program will crash if the user provides non-monotonically-increasing timestamps to the Update and MakeSnapshot functions. It seems to me that it's fairly easy to adapt both of those methods to accept input that's out of order: Update can ignore such updates, and MakeSnapshot already just returns empty snapshots for out-of-range timestamps.

File: src/medida/stats/ckms_sample.cc [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+    cur_window_->insert(value);
+}
+
+Snapshot CKMSSample::Impl::MakeSnapshot(Clock::time_point timestamp) {
+    std::lock_guard<std::mutex> lock{mutex_};
+    AssertValidTime(timestamp);
+    if (IsInCurrentWindow(timestamp)) {
+        return {*prev_window_};
+    } else if (IsInNextWindow(timestamp)) {
+        return {*cur_window_};
```
Comment (graydon) [inline src/medida/stats/ckms_sample.cc]: Something about this feels off -- the whole point of having 2 windows was that you're never going to snapshot from the current window, only previous. This seems to be mixing up time-advancing and the choice of whether a given timestamp is inside a given window. 

I would like to propose a slightly different design for the way these methods work. 

  - Extract your "window advancing" code from lines 167-182 into its own method, call it `bool AdvanceWindows(Clock::time_point timestamp)`, and have it return false if `timestamp < cur_window_begin_`, else advance the windows as you're already doing and return true.
  - Remove `AssertValidTime` and `last_asserted_time_`. You're already tracking a monotonically-increasing timestamp in the form of `cur_window_begin_`.
  - Rewrite `Update` like `if (AdvanceWindows(timestamp)) { cur_window_->insert(value); }`
  - Rewrite `MakeSnapshot` like `if (AdvanceWindows(timestamp)) { return {*prev_window_}; } else { return {CKMS()} }`

File: src/medida/stats/snapshot.cc [hunk 3]
```diff
@@ -22,24 +22,55 @@ static const double kP999_Q = 0.999;
 
 class Snapshot::Impl {
  public:
-  Impl(const std::vector<double>& values);
-  ~Impl();
-  std::size_t size() const;
-  double getValue(double quantile) const;
-  double getMedian() const;
-  double get75thPercentile() const;
```
Comment (graydon) [inline src/medida/stats/snapshot.cc:30]: Why are all these const methods no longer const?

File: src/medida/stats/snapshot.cc [hunk 4]
```diff
@@ -22,24 +22,55 @@ static const double kP999_Q = 0.999;
...
-  std::vector<double> getValues() const;
+  virtual std::size_t size() const = 0;
+  virtual double getValue(double quantile) = 0;
+  virtual double getMedian();
+  virtual double get75thPercentile();
+  virtual double get95thPercentile();
+  virtual double get98thPercentile();
+  virtual double get99thPercentile();
+  virtual double get999thPercentile();
+  virtual double sum() const { return 0; };
```
Comment (graydon) [inline src/medida/stats/snapshot.cc]: `sum`, `min` and `max` should probably be pure virtual like `getValue`; then you can provide implementations of them in `VectorImpl` as well as `CKMSImpl`.

File: src/medida/histogram.cc [hunk 5]
```diff
@@ -146,18 +150,21 @@ void Histogram::Impl::Clear() {
 
 std::uint64_t Histogram::Impl::count() const {
   std::lock_guard<std::mutex> lock {mutex_};
-  return count_;
+  return sample_type_ == kCKMS ? sample_->MakeSnapshot().size() : count_;
```
Comment (graydon) [inline src/medida/histogram.cc]: These methods that switch behaviour based on `sample_type_ == kCKMS` aren't quite right, I think. I want to take a step back and describe what we're trying to do, what we have on hand, and what our choices are / should be.

The histogram is incrementally tracking the following "all time" metrics (separate from the sample contents):

  - count
  - sum
  - min
  - max
  - mean
  - std dev

For each of these we can calculate a per-sample-snapshot value instead of an all-time value, using the `sample_->MakeSnapshot().{min,max,...}()` methods. The question is: should we? And if so, where?

Currently the methods on the histogram report the all-time values, in all cases. `Histogram::impl::count` returns the all-time `_count`. I think this is probably the way we want to keep it -- it's the only way to get that value from outside the class.

What I think we might want to change is what we _report_ to prometheus (or report to JSON blobs that the prometheus exporter then translates to prometheus). In fact, if you look at the _reporting_ methods (for example `JsonReporter::Impl::Process`) you'll see that it already manually extracts the sample snapshot, and reports a _mix_ of some all-time values (count, sum, min, max, mean, and stddev) and the rest (the percentiles) from the sample snapshot.

So I think -- reading the prometheus page on [summary metrics](https://prometheus.io/docs/concepts/metric_types/#summary) -- that what we want to do here is:

  - Continue reporting `sum` and `count` as all-time metrics -- as prometheus _defines_ those as "all time" (see the link above).
  - Similarly leave `min`, `max`, `mean` and `stddev` as all-time; we don't actually forward them to prometheus so there's no point changing or breaking them.
  - Add one new metric called "100%" which reports the snapshot max, not the all-time max. This is the signal we _will_ forward to prometheus, to feed into max-over-time(....) max-of-max calculations for spike-detection.

This is the least overall work to add, changes the meaning of no existing metrics, and instead just adds the one new signal we want (per-sample-snapshot maximum) in a way that's notationally-consistent with existing per-sample-snapshot percentile reporting.

(Ideally the metrics could differentiate _in their names_ whether they're per-sample-snapshot or all-time, but I think it would be quite a lot of breakage to change all the metric names everywhere from eg. `count` to `all_time_count` and so forth. We'd have to change a ton of prometheus code to compensate, and we'd be breaking anyone who scrapes core metrics manually, so I am leaning towards a "least breaking" approach. Especially since we might just remove all this in another six months and shift to the prometheus client library.)

You'll want to add this new `100%` metric to each of the following reporting methods:

  - `JsonReporter::Impl::Process(Histogram& histogram)`
  - `JsonReporter::Impl::Process(Timer& timer)`
  - `ConsoleReporter::Impl::Process(Histogram& histogram)`
  - `ConsoleReporter::Impl::Process(Timer& timer)`

It's just 1 line of code, should be fairly easy. Then remove all these hunks that change Histogram behaviour depending on whether sample type is kCKMS.

I'm sorry for the runaround on this, I suspect I suggested what you've done here in our last conversation about it, but rethinking I think we can get away with less overall variation in behaviour / fewer special cases.

File: src/medida/timer.cc [hunk 6]
```diff
@@ -183,27 +189,27 @@ std::uint64_t Timer::Impl::count() const {
 
 
 double Timer::Impl::min() const {
-  return histogram_.min() / duration_unit_nanos_;
+  return histogram_.min();
```
Comment (graydon) [inline src/medida/timer.cc]: Why are you removing the time normalization here? This doesn't seem right.

File: src/medida/timer.cc [hunk 7]
```diff
@@ -245,20 +251,14 @@ void Timer::Impl::Clear() {
 void Timer::Impl::Update(std::chrono::nanoseconds duration) {
   auto count = duration.count();
   if (count >= 0) {
-    histogram_.Update(count);
+    histogram_.Update(count / duration_unit_nanos_);
```
Comment (graydon) [inline src/medida/timer.cc]: Ok, if I understand correctly you're just trying to avoid the way `GetSnapshot` assumes it's dealing with a vector-of-doubles sample by moving time-normalization ahead of sampling -- but I think this is dangerous since it risks increasing error, no? Like if `duration_unit_nanos_` is 1 millisecond (common) and we have a lot of events in the range of 1-2ms, this will quantize them all to the same value (1ms) with no statistical distribution (and worse: sub-millisecond events all just become zero!)

The key here is that the methods on timer return _doubles_ at the duration unit resolution. So it's possible for `Timer::min` or `Timer::mean` to return a fractional millisecond, when configured with millisecond duration unit. The division happening in those methods is double-precision devision. The division you're doing here in `histogram_.Update(count / duration_unit_nanos_)` is integer division, which just truncates-away the fractional part.

I think a way to get where you're going might be something more like:
  - modify the `GetSnapshot` method to `GetSnapshot(uint64_t divisor=1)`
  - Pass the divisor through to the constructors of `Snapshot::VectorImpl` and `CKMSImpl`
  - In the `VectorImpl` case, do the vector-normalization loop that's currently in `Timer::Impl::GetSnapshot`
  - In the `CKMSImpl` case, just store the divisor as a member of `CKMSImpl` and then use it when calculating normalized answers to all the public accessors (`getValue`, `variance`, `min`, `max`, `sum`)
  - Make sure all the division you're doing in both of those cases (`VectorImpl` and `CKMSImpl`) is double division. It _should_ be, but be careful to be sure!)

File: test/test_timer.cc [hunk 8]
```diff
@@ -56,36 +59,39 @@ TEST_F(TimerTest, aBlankTimer) {
...
-  timer.Update(std::chrono::milliseconds(20));
-  std::this_thread::sleep_for(std::chrono::milliseconds(500));
-  timer.Update(std::chrono::milliseconds(20));
-  std::this_thread::sleep_for(std::chrono::milliseconds(500));
-  timer.Update(std::chrono::milliseconds(30));
-  std::this_thread::sleep_for(std::chrono::milliseconds(500));
-  timer.Update(std::chrono::milliseconds(40));
-
-  EXPECT_EQ(5, timer.count());
+  // As CKMS isn't very accurate when there's very few samples,
```
Comment (graydon) [inline test/test_timer.cc:62]: This test should also use timer events that are much shorter than 1ms (or whatever the duration unit is) so that you can test to see that there's no additional quantization error.

**Review by graydon — CHANGES_REQUESTED** [review]
Good first pass! Still a few things I think need work -- comments inline.

File: src/medida/histogram.cc (hunk 5 above)
Comment (hidenori-shinohara) [reply src/medida/histogram.cc]: That makes a lot of sense. Initially, I thought that it might not hurt to report `min`, `max`, `mean` and `stddev` based on the previous time window, instead of all-time history, but I realized that it'd be really confusing to report some metrics based on all-time history and some metrics based on partial history.

I'll implement this by:

- Adding `max` to the `Snapshot` class since `CKMS` doesn't seem to be 100% accurate when it comes to the maximum value. Then I'll report the `max` value from the snapshot as `100%` in those 4 reporting methods.
- Removing the `min` `sum` `variance` that I added to the `Snapshot` class as it's no longer necessary.
- Similarly, I'll remove `min` `sum` `variance` from `CKMS`.

This would simplify the implementation quite a bit. Also, I think this helps with migrating the change in Core. Some tests in Core seem to depend on the all-time `count` metric, and changing it to use the 30-second window would have probably broken those tests.

@MonsieurNicolas Just wanted to mention you here to keep you in the loop.

File: src/medida/timer.cc (hunk 7 above)
Comment (hidenori-shinohara) [reply src/medida/timer.cc]: I didn't notice that `Histogram`'s `Update` function takes an integer. I thought it took doubles since functions like `min`, `max` and `sum` are all doubles. Will update it!

File: src/medida/stats/ckms_sample.cc (hunk 1 above)
Comment (hidenori-shinohara) [reply src/medida/stats/ckms_sample.cc]: Makes sense! Updated it.

File: src/medida/stats/ckms_sample.cc (hunk 2 above)
Comment (hidenori-shinohara) [reply src/medida/stats/ckms_sample.cc]: Fixed.

File: src/medida/stats/snapshot.cc (hunk 3 above)
Comment (hidenori-shinohara) [reply src/medida/stats/snapshot.cc:30]: We chatted offline, but for future reference: I removed `const` since `CKMSSample` has a member `CKMS` which updates its state when calculating the quantile. But I learned that I can just have a pointer to `CKMS`, so I reverted the change (now they are back to being `const`).

File: src/medida/stats/snapshot.cc (hunk 4 above)
Comment (hidenori-shinohara) [reply src/medida/stats/snapshot.cc]: I removed `sum` and `min` based on following discussions, but I made `max` pure virtual.

File: src/medida/histogram.cc (hunk 5 above)
Comment (hidenori-shinohara) [reply src/medida/histogram.cc]: Done. This simplified the change quite a bit :)

File: src/medida/timer.cc (hunk 6 above)
Comment (hidenori-shinohara) [reply src/medida/timer.cc]: Reverted

File: src/medida/timer.cc (hunk 7 above)
Comment (hidenori-shinohara) [reply src/medida/timer.cc]: Done! It's slightly annoying that we'll end up with an unused variable in `src/medida/timer.cc`, but I think it's okay?:

```
stats::Snapshot Timer::Impl::GetSnapshot(uint64_t divisor) const {
  // Timer should always use duration_unit_nanos_ as the divisor
  // since that is the only thing that makes sense.
  // However, Snapshot class inherits GetSnapshot from SamplingInterface
  // which forces GetSnapshot to have an argument.
  // Therefore, we will ignore the divisor argument, and pass duration_unit_nanos_.
  return histogram_.GetSnapshot(duration_unit_nanos_);
}
```

File: test/test_timer.cc (hunk 8 above)
Comment (hidenori-shinohara) [reply test/test_timer.cc:62]: I added a test that uses 1 nano second to see if the division is handled correctly.

File: src/medida/histogram.cc [hunk 9]
```diff
@@ -203,8 +206,8 @@ double Histogram::Impl::variance() const {
 }
 
 
-stats::Snapshot Histogram::Impl::GetSnapshot() const {
-  return sample_->MakeSnapshot();
+stats::Snapshot Histogram::Impl::GetSnapshot(uint64_t divisor = 1) const {
```
Comment (graydon) [inline src/medida/histogram.cc]: This isn't quite right, but close. There are two problems:

  - You're providing the default argument to the `Histogram::Impl::GetSnapshot` class member, not the `Histogram::GetSnapshot` class member. The point of the default is to make _clients_ of `Histogram` not have to pass an argument if they don't want to (eg. when someone is using a Histogram outside of a Timer).
  - You're providing the default argument in the out-of-line member function _definition_, where you should be providing it in the _declaration_ (that is: in the `class { ... }` block). I actually didn't know the thing you're doing here works at all! My first thought was that it would not compile. It turns out it's legal but works in a _very odd way_ that's TU-specific -- the rules are [quite subtle](https://en.cppreference.com/w/cpp/language/default_arguments) and I had no idea this special case existed, but .. let's not use it.

In other words, we're going to want:

  - `Histogram::GetSnapshot(uint64_t divisor = 1)` declared inside the `class Histogram` declaration
  - No other default arguments mentioned anywhere else -- not in the `Impl` class and not in any of the out-of-line method definitions. Just that one declaration site.

**EDIT**: Actually now that I see (in comments below) that `GetSnapshot` comes from `SamplingInterface`, you're going to need two separate methods to solve that (the method on `SamplingInterface` shouldn't have a divisor argument at all, only an additional method on `Histogram`) so I think actually you will wind up with _no_ functions that use default arguments. Sorry for the digression.

File: src/medida/timer.cc [hunk 10]
```diff
@@ -251,14 +257,13 @@ void Timer::Impl::Update(std::chrono::nanoseconds duration) {
...
 
-stats::Snapshot Timer::Impl::GetSnapshot() const {
-  auto values = histogram_.GetSnapshot().getValues();
-  std::vector<double> converted;
-  converted.reserve(values.size());
-  for (auto& v : values) {
-    converted.push_back(v / (double)duration_unit_nanos_);
-  }
-  return {converted};
+stats::Snapshot Timer::Impl::GetSnapshot(uint64_t divisor) const {
```
Comment (graydon) [inline src/medida/timer.cc]: Oh no, this is a mess! I do not like this at all. Sorry, I didn't realize `GetSnapshot` came from `SamplingInterface` as well. Here is my recommended approach:

  - Leave `SamplingInterface::GetSnapshot()` as-is: do not add a `divisor` argument to it.
  - Leave the signature of `Timer::GetSnapshot` and `Timer::Impl::GetSnapshot` alone too: no new `divisor` argument.
  - Have _two_ methods on `Histogram`: both `Histogram::GetSnapshot() const override` which is Histogram's implementation of `SamplingInterface::GetSnapshot() const` as well as an _additional_ method `Histogram::GetSnapshot(uint64_t divisor) const` which is a method _specific_ to `Histogram`. Same name but otherwise no relationship to the method on `SamplingInterface`. Have the `Histogram::GetSnapshot()` method call `Histogram::GetSnapshot(1)`
  - Have `Timer::Impl::GetSnapshot()` call `histogram_.GetSnapshot(duration_unit_nanos_.count())` here like you're doing. Timer knows it has a `Histogram`, not just any old `SamplingInterface`, so it can call the `Histogram`-specific method that takes a `divisor` on its `histogram_` member.

File: test/test_timer.cc (hunk 8 above)
Comment (graydon) [reply test/test_timer.cc:62]: Great.

File: src/medida/stats/ckms_sample.h [hunk 11]
```diff
@@ -0,0 +1,63 @@
...
+// Each of size, Update, and MakeSnapshot has two versions, and
+// the one without a timestamp calls the other one with the current time.
+//
+// Unless there's a good reason to do so,
+// you should always use the one WITHOUT the timestamp.
+//
+// The one with a timestamp is generally used for testing. We pass a timestamp
+// as a way to fast-forward time to make testing easier.
+//
+// Regardless of which ones you use, the only rule that the caller must follow is
```
Comment (graydon) [inline src/medida/stats/ckms_sample.h:36]: This paragraph of this comment is no longer relevant, should be removed.

File: src/medida/stats/snapshot.cc (hunk 3 above)
Comment (graydon) [reply src/medida/stats/snapshot.cc:30]: Thanks.

**Review by graydon — CHANGES_REQUESTED** [review]
Getting there! I think just one more round of changes ought to do it. Mostly just fiddling with the routing of default arguments.

File: src/medida/stats/ckms_sample.h (hunk 11 above)
Comment (hidenori-shinohara) [reply src/medida/stats/ckms_sample.h:36]: I think it's still relevant. The class no longer throws when the caller goes back in time, but not going back in time is (I believe) the only simple way for the caller to not mess up. Going back in time by any amount, no matter how small, always has the risk of ending up in the wrong window.

File: src/medida/histogram.cc (hunk 9 above)
Comment (hidenori-shinohara) [reply src/medida/histogram.cc]: (I made the changes in the next comment instead of this)

File: src/medida/timer.cc (hunk 10 above)
Comment (hidenori-shinohara) [reply src/medida/timer.cc]: Aha! This made the code much less messier :) Thanks!

File: src/medida/histogram.cc (hunk 5 above)
Comment (hidenori-shinohara) [reply src/medida/histogram.cc]: For some reason, I completely forgot to update the reporting functions. I just did and pushed those changes here.

File: src/medida/timer.cc (hunk 7 above)
Comment (graydon) [reply src/medida/timer.cc]: (This was fixed differently -- resolved)

File: src/medida/stats/ckms_sample.h (hunk 11 above)
Comment (graydon) [reply src/medida/stats/ckms_sample.h:36]: ok!

**Comment by graydon** [comment]
I think this is good now, merging!
